refacor: Обновление аннотации @EqualsAndHashCode в классе ExtendedAd

linskay opened the pull request:
- Убедились, что аннотация @EqualsAndHashCode(callSuper = true) корректно учитывает поля родительского класса.

linskay committed:
refacor: Обновление аннотации @EqualsAndHashCode в классе ExtendedAd

         - Убедились, что аннотация @EqualsAndHashCode(callSuper = true) корректно учитывает поля родительского класса.

linskay committed:
refactor: рефакторинг логики обновления объявления и создание класса ErrorMessages

Описание:

Проведен рефакторинг логики обновления объявления для улучшения читаемости и поддержки кода.

Создан класс ErrorMessages для хранения всех сообщений об ошибках в одном месте.

Улучшено логирование в методах, связанных с созданием и обновлением объявлений.

Добавлены проверки на null и пустые значения для входных данных.

Известная проблема: баг с созданием объявления через метод addAd остался неисправленным.

linskay committed:
refactor+feat: Настроена конфигурация, аутентификация и авторизация

              - Добавила конфигурацию Spring Security с использованием JdbcUserDetailsManager.
              - Настроила таблицы `users` и `authorities` для хранения пользователей и их ролей.
              - Реализовала регистрацию пользователей с сохранением в таблицы `users` и `authorities.
              - Добавила возможность получения username текущего пользователя через объект Authentication.
              - Данные сохраняются корректно и могут быть использованы для аутентификации и авторизации.

linskay committed:
refactor+feat: Реализованы методы для работы с объявлениями

               - Перенесла логику из контроллера в сервис метода получения всех объявлений (`/ads`), доступный всем пользователям.
               - Добавила метод получения объявления по ID (`/ads/{id}`), доступный только авторизованным пользователям.
               - Настроила Spring Security для ограничения доступа к `/ads/{id}`.
               - Добавила аннотацию `@PreAuthorize("isAuthenticated()")` для метода `getAd`.
               - Реализовала возврат DTO (`Ads` и `ExtendedAd`).
               - Добавила обработку ошибок для случая, когда объявление не найдено.

linskay committed:
Обновлены миграции таблицы users и авторизации:
- Убрано поле username в таблице users.
- Присвоено поле email как уникальный идентификатор.

Обновлены DTO для аутентификации:
- Рефакторинг DTO для регистрации, логина и авторизации.

Рефакторинг сервиса аутентификации:
- Обновлены методы регистрации, логина и авторизации.

Рефакторинг сервиса пользователя:
- Добавлены методы:
  - Обновить данные пользователя (updateUser).
  - Обновить пароль пользователя (updatePassword) (возможен баг).
  - Обновить аватар пользователя (updateUserImage).
  - Получить информацию о пользователе (getUserInfo).

Обновлен контроллер пользователя:
- Добавлены эндпоинты для обновления данных, пароля и аватара.
- Реализована загрузка и сохранение изображений в проекте (путь: src/main/resources/static/uploads/images).

linskay committed:
Обновлена миграция, рефактор сервиса

linskay committed:
feat+refactor: обновление и улучшение архитектуры сервисов и контроллеров

- Отрефакторен сервис авторизации (AuthServiceImpl): улучшена логика методов, добавлены осмысленные логи, устранены избыточные проверки.
- Переработан UserServiceImpl: все методы протестированы и работают корректно, добавлены аннотации для валидации, исправлена логика работы с пользователями.
- Добавлены аннотации в UserController для унификации API и улучшения читаемости кода.
- Исправлены миграции: добавлены необходимые поля и индексы для таблиц пользователей и объявлений.
- Созданы классы AuthenticationUtils и ImageUtils для вынесения повторяющейся логики в утилитарные методы.
- Реализован сервис объявлений (AdService) и соответствующий контроллер (AdController): добавлены методы для создания, обновления, удаления и получения объявлений.
- Обновлены настройки application.properties: добавлены пути для хранения и доступа к изображениям.
- Изменены маппинги (Mappings):
  - Обновлены DTO и Entity маппинги для корректной работы с изображениями, комментариями и связанными сущностями.
  - Добавлены новые маппинги для поддержки дополнительных полей в объявлениях и пользователях.

linskay committed:
feat: добавление функционала для работы с комментариями
   - Реализован метод `addComment` для добавления нового комментария к объявлению.
   - Реализован метод `updateComment` для обновления текста существующего комментария.
   - Реализован метод `deleteComment` для удаления комментария.
   - Реализован метод `getCommentsByAdId` для получения всех комментариев объявления.
   - Добавлена проверка прав доступа пользователя для операций обновления и удаления комментариев.
   - Добавлен эндпоинт `POST /ads/{id}/comments` для создания нового комментария.
   - Добавлен эндпоинт `PATCH /ads/{adId}/comments/{commentId}` для обновления комментария.
   - Добавлен эндпоинт `DELETE /ads/{adId}/comments/{commentId}` для удаления комментария.
   - Добавлен эндпоинт `GET /ads/{id}/comments` для получения списка комментариев объявления.
   - Добавлена аннотация `@PreAuthorize` для проверки аутентификации пользователя.
   - Добавлен маппер `CommentMapper` для преобразования сущностей (`CommentEntity`) в DTO (`Comment`) и обратно.
   - Использован метод `mapInstantToEpochMillis` для преобразования времени создания комментария в миллисекунды.
   - Добавлены кастомные исключения: `CommentNotFoundException`, `AccessDeniedException`, `AdNotFoundException`.
   - Настроена глобальная обработка исключений в `GlobalExceptionHandler`.
   - Добавлен метод `findByAdEntity` в интерфейс `CommentRepository` для поиска комментариев по объявлению.
   - Добавлен метод `findByAdId` с использованием аннотации `@Query` для поиска комментариев по ID объявления.
   - Добавлены аннотации `@Schema` для описания полей в OpenAPI.
   - Добавлены логи для отслеживания выполнения операций с комментариями (добавление, обновление, удаление, получение).
   - Убедились, что все эндпоинты корректно обрабатывают запросы и возвращают ожидаемые ответы.
   - Устранена ошибка `StackOverflowError`, вызванная циклическими ссылками между сущностями.
   - Устранена ошибка `MissingPathVariableException`, связанная с некорректным форматом URL.
   - Устранена ошибка `cannot find symbol`, вызванная отсутствием метода `builder()` в классе `Comments`.
    - Добавлены аннотации `@JsonBackReference` и `@JsonManagedReference` для предотвращения циклических ссылок при сериализации JSON.

linskay committed:
test: добавлены unit-тесты для мапперов

Добавлены unit-тесты для проверки корректности работы мапперов:
- UserMapper: тестирование преобразования между UserEntity, User, Login, Register и UpdateUser.
- AdMapper: тестирование преобразования между AdEntity, Ad и ExtendedAd.
- CommentMapper: тестирование преобразования между CommentEntity и Comment.

Тесты покрывают основные сценарии:
- Преобразование сущностей в DTO и обратно.
- Проверка игнорируемых полей (например, id, createdAt).
- Проверка корректности маппинга связанных объектов (например, author, comments).

Используются библиотеки AssertJ и JUnit 5 для написания тестов. Добавлены аннотации @DisplayName для описания каждого тестового случая.

linskay committed:
refactor: рефакторинг проекта и добавление новых функций

1. Рефакторинг pom.xml:
   - Добавлена группировка зависимостей для улучшения читаемости:
     - Spring Boot Starter зависимости
     - База данных (PostgreSQL, Liquibase)
     - Тестирование (Spring Boot Test, Spring Security Test)
     - Документация API (Springdoc OpenAPI)
     - Lombok и MapStruct
   - Обновлена версия PostgreSQL до 42.7.2 для устранения критической уязвимости безопасности.

2. Добавлены тесты:
   - Написаны unit-тесты для DTO `Login` и `Register` с проверкой всех случаев валидации:
     - Проверка аннотаций `@NotBlank`, `@Size`, `@Email`, `@Pattern`, `@NotNull`.
   - Тесты покрывают все возможные ошибки валидации и корректные сценарии.

3. Переписаны сервис и контроллер аутентификации:
   - Реализованы изменения в AuthController и AuthService для соответствия спецификации:
     - Добавлена валидация входных данных из DTO на уровне контроллера.
     - Внедрены кастомные ошибки с понятными текстовыми сообщениями для пользователя.
   - Улучшена обработка исключений через глобальный обработчик ошибок.

4. Валидация данных:
   - Добавлена валидация входных данных на уровне контроллера с использованием Hibernate Validator

5. Добавлен UserDetailsAdapter

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/util/ImageUtils.java:
```diff
@@ -0,0 +1,63 @@
+    public byte[] getImageAsBytes(String imagePath) {
+        try {
+            // Удаляем базовый URL, чтобы получить только путь к файлу
+            String relativePath = imagePath.replace(baseUrl, "");
```
replace тяжелая операция, лучше тогда substring, а еще лучше обрабатывать вообще без директории и инкапсулировать её внутри приложения. Наружу отдавать только название принимать тоже, остальные операции делать тут. Исправлять не обязательно

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/util/ImageUtils.java:
```diff
@@ -0,0 +1,63 @@
+
+            return Files.readAllBytes(filePath);
+        } catch (IOException e) {
+            throw new RuntimeException("Не удалось прочитать изображение", e);
```
RuntimeException не кидаем, только кастомные

AleksandrZhukovJava commented on src/test/java/ru/skypro/homework/dto/LoginTest.java:
```diff
@@ -0,0 +1,163 @@
+    @Test
+    @DisplayName("Проверка корректного создания объекта Login")
+    void testValidLogin() {
+        try (var validatorFactory = Validation.buildDefaultValidatorFactory()) {
```
без var, тут и далее

linskay commented on pom.xml:
```diff
@@ -1,133 +1,136 @@
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>postgresql</artifactId>
+            <version>42.7.2</version>
```
спрингбут автоматом подтягивает версию 42.3.8, https://osv.dev/vulnerability/GHSA-24rp-q3w6-vc56. или, в целом без разницы и не влияет ни на что?🙃

linskay committed:
feat(config): разделена логика управления пользователями через JdbcUserDetailsConfig

- Создан отдельный конфигурационный класс JdbcUserDetailsConfig
- Исправлен маппинг данных для корректного отображения изображений пользователей и объявлений на фронте
- Улучшена читаемость кода за счет разделения ответственности между компонентами.

linskay committed:
feat: Настройка Spring Security и рефакторинг процесса аутентификации

Spring Security для защиты endpoint'ов приложения и реализует кастомный процесс аутентификации с использованием JdbcUserDetailsManager.

linskay committed:
ups. save.eat-sleep-coding-repeat

linskay committed:
tests dto

linskay committed:
refactor+test

linskay committed:
refactor:
exceptions
controllers

linskay committed:
ай нид хелп ююююю(((((((((((

AleksandrZhukovJava commented on pom.xml:
```diff
@@ -1,133 +1,146 @@
+            <artifactId>spring-security-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Это уже есть внутри spring-boot-starter-test

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/config/JdbcSecurityConfig.java:
```diff
@@ -0,0 +1,40 @@
+        return manager;
+    }
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
```
Это бин относящейся к общей структуре шифрования, а не к JDBC. Перенесите в общий конфиг безопасности

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/config/SwaggerConfig.java:
```diff
@@ -1,19 +1,40 @@
 @Configuration
 public class SwaggerConfig {
 
+    @Value("${app.name}")
+    private String appName;
+
+    @Value("${app.description}")
+    private String appDescription;
+
+    @Value("${app.version}")
+    private String appVersion;
```
Можно изучить @ConfigurationProperties()
тогда сможете доставать сразу все поля

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -2,61 +2,59 @@
+                .authorizeHttpRequests((authz) -> authz
+                        .mvcMatchers(AUTH_WHITELIST).permitAll()
+                        .mvcMatchers("/ads/**", "/users/**", "/images/**").authenticated()
+                        .anyRequest().denyAll()
```
заставляете каждого нового разработчика на вашем проекте пройти мучительный путь осознания почему его новый ендпоинт отдает ошибку что бы он не делал. Лучше сделайте все authenticated

linskay replied:
)

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -2,61 +2,59 @@
     @Bean
-    public PasswordEncoder passwordEncoder() {
-        return new BCryptPasswordEncoder();
+    public UserDetailsService userDetailsService(ru.skypro.homework.repository.UserRepository userRepository) {
+        return email -> userRepository.findByEmail(email)
+                .map(UserDetailsAdapter::new)
+                .orElseThrow(() -> new org.springframework.security.core.userdetails.UsernameNotFoundException("User not found with email: " + email));
```
У вас уже есть JDBC оставьте что-то одно

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -2,61 +2,59 @@
     }
 
-}
+    @Bean
+    public AuthenticationManager authenticationManager(HttpSecurity http, BCryptPasswordEncoder bCryptPasswordEncoder, UserDetailsService userDetailsService)
+            throws Exception {
+        return http.getSharedObject(AuthenticationManagerBuilder.class)
+                .userDetailsService(userDetailsService)
+                .passwordEncoder(bCryptPasswordEncoder)
+                .and()
+                .build();
+    }
+}
```
Он сам это конфигурирует, в ручную не нужно

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/controller/ImagesController.java:
```diff
@@ -0,0 +1,37 @@
+public class ImagesController {
+    ImageUtils imageUtils;
+
+    @SneakyThrows
+    @GetMapping("{imagePath}")
+    @PreAuthorize("isAuthenticated()")
+    public ResponseEntity<byte[]> image(Object o) throws IOException {
+        return image(null);
+    }
```
что-то предназначенное что бы подвесить запрос в бесконечную рекурсию я верно понял замысел?)

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/controller/UserController.java:
```diff
@@ -3,94 +3,119 @@
     }
 
+
+//    @PatchMapping(value = "/me/image", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+//    @Operation(
+//            summary = "Обновление аватара авторизованного пользователя",
+//            responses = {
+//                    @ApiResponse(responseCode = "200", description = "OK"),
+//                    @ApiResponse(responseCode = "401", description = "Не авторизован")
+//            }
+//    )
+//    @PreAuthorize("isAuthenticated()")
+//    public ResponseEntity<String> updateUserImage(
+//            @RequestParam("image") MultipartFile image
+//    ) {
+//        String imageUrl = Arrays.toString(userService.updateUserAvatar(image));
+//
+//        return ResponseEntity.ok().body(imageUrl);
+//    }
```
В финальной работе закомментированного кода быть не должно

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/AuthService.java:
```diff
@@ -1,9 +1,47 @@
+    @Transactional
+    public Long register(Register register) {
+        if (userRepository.existsByEmail(register.getUsername())) {
+            return null;
```
Лучше из методов никогда не возвращать null, тут больше подойдет ошибка

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -1,165 +1,111 @@
-            throw new ImageUploadException("Ошибка при загрузке изображения", e);
-        }
+    public Ads getAllAds() {
+        List<AdEntity> adEntities = adRepository.findAll();
```
Главное не делайте так в настоящем приложении, можете вытянуть 432958598 записей

linskay committed:
feat: Переработана аутентификация и обновлена конфигурация Swagger

- Аутентификация:
  - Переработана конфигурация Spring Security для использования JdbcUserDetailsManager
  - Устранены циклические зависимости в конфигурации бинов, возникавшие при использовании JdbcUserDetailsManager.
  - Контроллер аутентификации обновлен для использования AuthenticationManager для проверки учетных данных и установки Authentication в SecurityContextHolder.

- Swagger:
  - Обновлена конфигурация Swagger.

- Зависимости:
  - Удалена лишняя зависимость.

  -Обновлена обработка изображений

linskay committed:
исправлена картинка

linskay committed:
bug-fix:
иcправлена авторизация, аутентификация, настроены кеш-куки, хранение в сессии;
картинки видит фронт!!!

linskay committed:
ревью

linskay committed:
bug-fix:
- исправлена логика получения картинок
- исправлен метод обновления пароля
- исправлен конфиг секьюрити(рефактор)

linskay committed:
fix:
-добавлены тестовые пользователи
- исправлен конфиг

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/config/TestUsersConfig.java:
```diff
@@ -0,0 +1,46 @@
+import ru.skypro.homework.model.UserEntity;
+import ru.skypro.homework.repository.UserRepository;
+
+@Configuration
```
Тогда @Profile("dev") в проде так нельзя делать

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -2,61 +2,104 @@
+        return source;
+    }
+
+    @Bean
+    public JdbcUserDetailsManager userDetailsManager(DataSource dataSource, PasswordEncoder passwordEncoder) {
+        JdbcUserDetailsManager manager = new JdbcUserDetailsManager(dataSource);
+
+            manager.setUsersByUsernameQuery("SELECT email, password, enabled FROM users WHERE email = ?");
+            manager.setAuthoritiesByUsernameQuery("SELECT email, authority FROM authorities WHERE email = ?");
+            manager.setCreateAuthoritySql("INSERT INTO authorities (email, authority) VALUES (?, ?)");
+            manager.setCreateUserSql("INSERT INTO users (email, password, enabled, first_name, last_name, phone) VALUES (?, ?, ?, ?, ?, ?)");
+            manager.setDeleteUserSql("DELETE FROM users WHERE email = ?");
+            manager.setUpdateUserSql("UPDATE users SET password = ?, enabled = ? WHERE email = ?");
+            manager.setDeleteUserAuthoritiesSql("DELETE FROM authorities WHERE email = ?");
+            return manager;
+        }
```
Это отдельный класс компонент, такие бины нужны когда нужна какая то кастомизация, а тут достаточно @Component

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/controller/GlobalExceptionHandler.java:
```diff
@@ -1,71 +1,81 @@
-        ex.getBindingResult().getFieldErrors().forEach(error ->
-                errors.put(error.getField(), error.getDefaultMessage()));
-        return ResponseEntity.badRequest().body(errors);
+    @ExceptionHandler(EntityNotFoundException.class)
+    public ResponseEntity<String> handleEntityNotFound(EntityNotFoundException e) {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Объект не найден: " + e.getMessage());
```
Тут все можно заменить на 1 хэндлер общего родителя всех этих исключений, там же прокидывать в ответ статус который из него же и получать

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/controller/ImagesController.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    private String determineContentType(String filename) {
+        if (filename.toLowerCase().endsWith(".png")) {
+            return MediaType.IMAGE_PNG_VALUE;
+        }
+        if (filename.toLowerCase().endsWith(".gif")) {
+            return MediaType.IMAGE_GIF_VALUE;
+        }
+        return MediaType.IMAGE_JPEG_VALUE; // по умолчанию
+    }
```
Логичнее 3 кейса и ошибка по дефолту

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/AuthService.java:
```diff
@@ -1,9 +1,48 @@
+    @Transactional
+    public Long register(Register register) {
+        if (userRepository.existsByEmail(register.getUsername())) {
+            return null;
```
Лучше из методов никогда не возвращать null, тут больше подойдет ошибка (х2)

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/AuthService.java:
```diff
@@ -1,9 +1,48 @@
+
+        UserEntity userEntity = userMapper.toUser(register);
+        userEntity.setPassword(passwordEncoder.encode(register.getPassword()));
+        userEntity.setImage("/images/default-avatar.jpg");
```
в проперти

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/AuthService.java:
```diff
@@ -1,9 +1,48 @@
+        UserDetails userDetails = User.builder()
+                .username(register.getUsername())
+                .password(userEntity.getPassword())
+                .roles(String.valueOf(register.getRole()))
```
register.getRole().name

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -1,165 +1,119 @@
+                        AdNotFoundException(id));
+        UserEntity currentUser = getCurrentUser();
+        if (!adEntity.getAuthor().equals(currentUser)) {
+            throw new AccessDeniedException("Нет прав на просмотр этого объявления");
```
В коде сервисов вообще нигде не должно быть String переменных вживую. Они все либо убраны внутрь классов которые их используют, либо инджектятся из пропертей или бандлов (кроме логов)

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/service/impl/UserServiceImpl.java:
```diff
@@ -2,141 +2,81 @@
+            try {
+                imageUtils.deleteImage(user.getImage());
+            } catch (IOException e) {
+                log.warn("Не удалось удалить старое изображение: {}", user.getImage());
```
Мне кажется это важная ошибка, лога не достаточно. Я бы закрывал юзеру функционал если такое произошло (исключение) что бы он шел в тех. под и нам приносили баг

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/util/ImageUtils.java:
```diff
@@ -0,0 +1,53 @@
+
+        Files.copy(file.getInputStream(), filePath, StandardCopyOption.REPLACE_EXISTING);
+
+        return "/images/" + fileName;
```
в проперти тут и далее

AleksandrZhukovJava commented on src/main/java/ru/skypro/homework/util/ImageUtils.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public void deleteImage(String filename) throws IOException {
+        Path filePath = Paths.get(uploadDirPath).resolve(filename.replace("/images/", ""));
```
лучше substring
